src/mlpack/methods/gmm/diagonal_constraint.hpp

```diff
@@ -30,6 +30,13 @@ class DiagonalConstraint
     covariance = arma::diagmat(arma::clamp(covariance.diag(), 1e-10, DBL_MAX));
   }
 
+  //! Force a covariance vector to be non-zero.
```

Comment:
Can you clarify the documentation here to point out that the parameter here is just the diagonal of the covariance matrix?  Also I think we need to make these changes to `EigenvalueRatioConstraint` and `PositiveDefiniteConstraint`.  `PositiveDefiniteConstraint` should be easy:

```
/**
 * Apply the positive definiteness constraint to the given diagonal covariance matrix
 * (which is represented as a vector), and ensure each value on the diagonal is at least 1e-50.
 */
static void ApplyConstraint(arma::vec& diagCovariance)
{
  // No need to eigendecompose, since the diagonal is already the eigenvalues.
  arma::uvec eigvalOrder = arma::sort_index(diagCovariance);
  arma::vec eigval = diagCovariance[eigvalOrder];

    // If the matrix is not positive definite or if the condition number is
    // large, we must project it back onto the cone of positive definite
    // matrices with reasonable condition number (I'm picking 1e5 here, not for
    // any particular reason).
    double maxEigval = -DBL_MAX;
    for (size_t i = 0; i < diagCovariance.n_elem; ++i)
      if (diagCovariance[i] > maxEigval)
        maxEigval = diagCovariance[i];

  for (size_t i = 0; i < diagCovariance.n_elem; ++i)
  {
    if ((diagCovariance[i] < 0.0) || ((maxEigval / diagCovariance[i]) > 1e5) ||
        (maxEigval < 1e-50))
    {
      diagCovariance[i] = std::max(maxEigval / 1e5, 1e-50);
    }
  }
}
```

For `EigenvalueRatioConstraint` it's a little more tricky but not much:

```
/**
 * Apply the eigenvalue ratio constraint to the given diagonal covariance matrix (represented as a vector).
 */
void ApplyConstraint(arma::vec& diagCovariance) const
{
  // The matrix is already eigendecomposed but we need to sort the elements.
  arma::uvec eigvalOrder = arma::sort_index(diagCovariance);
  arma::vec eigvals = diagCovariance(eigvalOrder);

  // Change the eigenvalues to what we are forcing them to be.  There
  // shouldn't be any negative eigenvalues anyway, so it doesn't matter if we
  // are suddenly forcing them to be positive.  If the first eigenvalue is
  // negative, well, there are going to be some problems later...
  eigvals = (eigvals[0] * ratios);

  // Reassemble the matrix.
  for (size_t i = 0; i < eigvalOrder.n_elem; ++i)
    diagCovariance[eigvalOrder[i]] = eigval[i];
}
```

Actually I also found a bug with `EigenvalueRatioConstraint`.  I'll have to post a fix in another PR.

Reply:
Oh! Thanks for clarification.

Reply:
Hey @rcurtin, in the `PositiveDefiniteConstraint`, do we need the `eigvalOrder` and `eigval`? I'm curious because it is not used in the code.

Reply:
You're right, I forgot to remove it in my draft.  I've updated the code above now.  :+1:

Reply:
That code is not guaranteed to be right so it's good that you double-checked it. :)

Reply:
Sure :)